dogechain-lab/dogechain
# [Txpool] Demote all pending transactions when nonce not right in building block

# Description

The `txpool` pending queues might not be able to recover back to normal when nonce too high.
The nonce gap might happen sometimes when block gossip slow down, and the new block arrived first.

This PR demotes all pending transactions, which might be un-promotable, re-broadcast them again, in order to include them in some other blocks.

It is only a patch, and do not solve the issue at all.
We should focus more on the next version `txpool` for more performance and security.
For now, high memory and CPU requirement is needed for the security.

# Changes include


# Checklist


## Testing


### Manual tests

* Set up a 4 validators network, a rpc node, with `--max-slots=4096` flag.
* Use some tools to generate huge amount transactions, some with `100 GWei` gas price, some with `50 GWei` gas price.
* Check the logs, we should see some `write miss some transactions with higher nonce` log after the `txpool is full`.

# Documentation update

Would update the documentation once review starts.
